### PR TITLE
Selector: Decode invalid escape code points to U+FFFD

### PR DESCRIPTION
**Pull request by rootvector2**
`unescapeSelector` decodes CSS hex escapes that the spec maps to U+FFFD wrong:

    \0       -> U+0000   (literal NUL)
    \D800    -> U+D800   (lone surrogate)
    \110000  -> garbage  (above the Unicode max)

Return U+FFFD for null, surrogate, and out-of-range escapes. Reachable via the seeded `.filter()` matcher; test added alongside the others.

### CODE REVIEW

**Comment by timmywil** [comment]
The workflow failures are unrelated. Chrome has been updated in CI and the failures we saw in beta related to `:enabled` and `:disabled` pseudos  have made it to Chrome stable.

File: src/selector/unescapeSelector.js [hunk 1]
```diff
@@ -5,21 +5,31 @@ import { whitespace } from "../var/whitespace.js";
...
 			// Strip the backslash prefix from a non-hex escape sequence
 			return nonHex;
 		}
 
+		// Per the CSS spec, a NULL, surrogate, or out-of-range code point is
+		// replaced with the REPLACEMENT CHARACTER (U+FFFD).
+		if ( codePoint === 0 || codePoint > 0x10FFFF ||
+			( codePoint >= 0xD800 && codePoint <= 0xDFFF ) ) {
+			return "\uFFFD";
+		}
```
Comment (gibson042) [inline src/selector/unescapeSelector.js]: gzippability improvements:
```suggestion
		var codePoint = "0x" + escape.slice( 1 ) - 0;

		if ( nonHex ) {

			// Strip the backslash prefix from a non-hex escape sequence
			return nonHex;
		}

		// Per the CSS spec, a NULL, surrogate, or out-of-range code point is
		// replaced with the REPLACEMENT CHARACTER (U+FFFD).
		// https://www.w3.org/TR/css-syntax-3/#consume-escaped-code-point
		if ( !codePoint || codePoint > 0x10FFFF ||
			( codePoint >= 0xD800 && codePoint < 0xE000 ) ) {
			return "\uFFFD";
		}
```

File: src/selector/unescapeSelector.js [hunk 2]
```diff
@@ -5,21 +5,31 @@ import { whitespace } from "../var/whitespace.js";
...
 		// surrogate pair
-		return high < 0 ?
-			String.fromCharCode( high + 0x10000 ) :
-			String.fromCharCode( high >> 10 | 0xD800, high & 0x3FF | 0xDC00 );
+		return codePoint > 0xFFFF ?
+			String.fromCharCode(
+				( codePoint - 0x10000 ) >> 10 | 0xD800,
+				( codePoint - 0x10000 ) & 0x3FF | 0xDC00
+			) :
+			String.fromCharCode( codePoint );
```
Comment (gibson042) [inline src/selector/unescapeSelector.js]: gzippability improvements:
```suggestion
		return codePoint < 0x10000 ?
			String.fromCharCode( codePoint ) :
			String.fromCharCode(
				( codePoint - 0x10000 ) >> 10 | 0xD800,
				( codePoint - 0x10000 ) & 0x3FF | 0xDC00
			);
```

File: test/unit/selector.js [hunk 3]
```diff
@@ -844,6 +844,26 @@ QUnit.test( "attributes - special characters", function( assert ) {
 		"Long numeric escape (non-BMP)" );
 } );
 
+QUnit.test( "attributes - invalid escaped code points", function( assert ) {
```
Comment (gibson042) [inline test/unit/selector.js:847]: Let's also include a test case with complete and off-by-one coverage, e.g. that `[data-attr='\0 \1 \D7FF \D800 \DFFF \E000 \10FFFF \110000']` matches an element with attribute value `"\uFFFD\u0001\uD7FF\uFFFD\uFFFD\uE000\uDBFF\uDFFF\uFFFD"`.

**Review by gibson042 — APPROVED** [review]
Thanks! This should be good to go after a few small tweaks.

File: src/selector/unescapeSelector.js (hunk 1 above)
Comment (rootvector2) [reply src/selector/unescapeSelector.js]: applied. went with the `"0x" + escape.slice( 1 ) - 0` form, `!codePoint`, and `< 0xE000`, and added the spec link.

File: src/selector/unescapeSelector.js (hunk 2 above)
Comment (rootvector2) [reply src/selector/unescapeSelector.js]: done, BMP branch first now.

File: test/unit/selector.js (hunk 3 above)
Comment (rootvector2) [reply test/unit/selector.js:847]: added it as a fifth assertion: seeded an element with value `�퟿��􏿿�` and matched it with the full `\0 \1 \D7FF \D800 \DFFF \E000 \10FFFF \110000` list, so both sides of each boundary are covered.

**Comment by timmywil** [comment]
Please rebase when you get a chance.

**Comment by rootvector2** [comment]
rebased onto main, clean. should be good now.

**Comment by rootvector2** [comment]
any update?
